svgdigitizer/electrochemistry/electrolyte.py

```diff
@@ -0,0 +1,196 @@
+        acids_bases = []
+        for i in self.electrolyte['components']:
+            if not i['type'] == 'solvent':
+                InChIKey = self.lookup_chemical_ids_from_name(i['name'])["InChIKey"]
+                temp = self.get_pKa(InChIKey)
+                for e in temp:
+                    q = i['concentration']['value'] * normalize_unit(i['concentration']['unit'])
+                    e.conc = q.to("mol / l").value
+                acids_bases.extend(temp)
+
+        system = System(*acids_bases)
+        system.pHsolve()
+
+        return system.pH
+
+    @classmethod
+    def lookup_CID_from_InChI(cls, InChIKey=None, InChI=None):
+        r"""
+        Lookup Pubchem CID from InChI or InChiKey.
+
+        EXAMPLES::
+
+            CID can be retrieved with InChIKey::
+            >>> from svgdigitizer.electrochemistry.electrolyte import Electrolyte
+            >>> Electrolyte.lookup_CID_from_InChI(InChIKey='NBIIXXVUZAFLBC-UHFFFAOYSA-N') # phosphoric acid
+            1004
+
+            CID can be retrieved with InChI::
+            >>> Electrolyte.lookup_CID_from_InChI(InChI='InChI=1S/H3O4P/c1-5(2,3)4/h(H3,1,2,3,4)') # phosphoric acid
+            1004
+
+        """
+        if InChIKey:
+            return pcp.get_cids(InChIKey, 'inchikey')[0]
+        elif InChI:
+            return pcp.get_cids(InChI, 'inchi')[0]
+
+    @classmethod
+    def lookup_chemical_ids_from_name(cls, chemical_name):
+        r"""
+        Lookup CID, InChI, and InChIKey from chemical name.
+
+        EXAMPLES::
+
+            CID can be retrieved with InChIKey::
+            >>> from svgdigitizer.electrochemistry.electrolyte import Electrolyte
+            >>> Electrolyte.lookup_chemical_ids_from_name('phosphoric acid')
+            {'CID': 1004, 'InChI': 'InChI=1S/H3O4P/c1-5(2,3)4/h(H3,1,2,3,4)', 'InChIKey': 'NBIIXXVUZAFLBC-UHFFFAOYSA-N'}
+
+        """
+        results = pcp.get_cids(chemical_name, 'name')
+        len_results = len(results)
+        if len_results > 1:
+            raise RuntimeError("Ambiguous chemical name! More than one chemical id return by pubchem. Use another chemical name")
+        elif len_results == 0:
+            raise RuntimeError(f"Name of chemical {chemical_name} not found in pubchem! Check spelling.")
+        else:
+            return pcp.get_properties(("InChIKey", "InChI"), results[0])[0]
+
+    def get_pKa(self, InChIKey):
+        r"""
+        Get pKa
+        Pubchem lookup only works for neutral acids!
+
+        """
+        pattern = re.compile(r'.*(?P<pKa>\d+\.\d*).*')
+        try:
+            return acid_base_dict[InChIKey]['components']
+        except KeyError:
+            cid = self.lookup_CID_from_InChI(InChIKey=InChIKey)
+            pKa_result = self.lookup_pKa(cid)['pKa']
+            # TODO support salts etc.
+            return [Acid(pKa=[pattern.match(i).group('pKa') for i in pKa_result.split(';')], charge=0)]
+
+    def lookup_pKa(self, cid):
```

Comment:
There seems to a problem with rate-limiting when asking a lot of questions to pubchem. That's going to be a problem when running this a lot in the website CI. But let's see if that's really the case first.

Reply:
@saraedum I think we are already hitting this limit as one of the tests failed due to `ServerBusy` error.
```Run pytest --doctest-modules svgdigitizer
============================= test session starts ==============================
platform darwin -- Python 3.10.0, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/runner/work/svgdigitizer/svgdigitizer
collected 54 items

svgdigitizer/__main__.py .....                                           [  9%]
svgdigitizer/svg.py ..............                                       [ 35%]
svgdigitizer/svgplot.py ...............                                  [ 62%]
svgdigitizer/electrochemistry/cv.py ............                         [ 85%]
svgdigitizer/electrochemistry/electrolyte.py .F...                       [ 94%]
svgdigitizer/electrochemistry/normalization.py .                         [ 96%]
svgdigitizer/test/cli.py ..                                              [100%]

=================================== FAILURES ===================================
_ [doctest] svgdigitizer.electrochemistry.electrolyte.Electrolyte._estimate_pH _
146 
147         Return an estimate of the pH of this aquoues electrolytes.
148 
149         EXAMPLES:
150 
151             >>> from svgdigitizer.electrochemistry.electrolyte import Electrolyte
152             >>> electrolyte_dict = {"type": "aq", "components": [{"name": "H2SO4","type": "acid", "concentration": {"value": 0.1, "unit": "M"}}, {"name": "Na2SO4","type": "salt", "concentration": {"value": 0.1, "unit": "M"}}]}
153             >>> electrolyte = Electrolyte(electrolyte_dict)
154             >>> electrolyte._estimate_pH()
UNEXPECTED EXCEPTION: PubChemHTTPError(<HTTPError 503: 'PUGREST.ServerBusy'>)
Traceback (most recent call last):
  File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/pubchempy.py", line 271, in request
    response = urlopen(apiurl, postdata)
  File "/usr/local/miniconda/envs/test/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/miniconda/envs/test/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/usr/local/miniconda/envs/test/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/usr/local/miniconda/envs/test/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/usr/local/miniconda/envs/test/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/usr/local/miniconda/envs/test/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 503: PUGREST.ServerBusy
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/miniconda/envs/test/lib/python3.10/doctest.py", line 1348, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest svgdigitizer.electrochemistry.electrolyte.Electrolyte._estimate_pH[3]>", line 1, in <module>
  File "/Users/runner/work/svgdigitizer/svgdigitizer/svgdigitizer/electrochemistry/electrolyte.py", line 166, in _estimate_pH
    InChI = Electrolyte.lookup_chemical_ids_from_name(i["name"])['InChI']
  File "/Users/runner/work/svgdigitizer/svgdigitizer/svgdigitizer/electrochemistry/electrolyte.py", line 206, in lookup_chemical_ids_from_name
    results = pcp.get_cids(chemical_name, "name")
  File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/pubchempy.py", line 424, in get_cids
    results = get_json(identifier, namespace, domain, 'cids', searchtype=searchtype, **kwargs)
  File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/pubchempy.py", line 299, in get_json
    return json.loads(get(identifier, namespace, domain, operation, 'JSON', searchtype, **kwargs).decode())
  File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/pubchempy.py", line 292, in get
    response = request(identifier, namespace, domain, operation, output, searchtype, **kwargs).read()
  File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/pubchempy.py", line 274, in request
    raise PubChemHTTPError(e)
pubchempy.PubChemHTTPError: 'PUGREST.ServerBusy'
/Users/runner/work/svgdigitizer/svgdigitizer/svgdigitizer/electrochemistry/electrolyte.py:154: UnexpectedException
=============================== warnings summary ===============================
../../../../../usr/local/miniconda/envs/test/lib/python3.10/site-packages/astropy/version.py:24
  /usr/local/miniconda/envs/test/lib/python3.10/site-packages/astropy/version.py:24: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.version import LooseVersion

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED svgdigitizer/electrochemistry/electrolyte.py::svgdigitizer.electrochemistry.electrolyte.Electrolyte._estimate_pH
=================== 1 failed, 53 passed, 1 warning in 38.90s ===================
Error: Process completed with exit code 1.```

Reply:
We might need some memoization on disk here. Have you tried out something like this https://stackoverflow.com/a/58947970/812379?